spec/controllers/users_controller_spec.rb

```diff
@@ -0,0 +1,170 @@
+    before do
+      get :liked_posts, params: { id: -1 }
+    end
+
+    it "sets a flash danger message" do
+      expect(flash[:danger]).to eq(I18n.t("user.not_found"))
+    end
+
+    it "redirects to the root path" do
+      expect(response).to redirect_to(root_path)
+    end
+  end
+
+  describe "GET #index" do
+    it "returns a successful response" do
+      get :index
+      expect(response).to be_successful
+    end
+
+    it "assigns @users and @user_items" do
+      get :index
+      expect(assigns(:users)).to match_array(users)
+      expect(assigns(:user_items)).to match_array(users.first(Settings.users.per_page))
+    end
+
+    it "paginates the users" do
+      get :index
+      pagy, user_items = assigns(:pagy), assigns(:user_items)
+      expect(pagy).to be_a(Pagy)
+      expect(user_items.size).to eq(Settings.users.per_page)
+    end
+  end
+
+  describe "GET #new" do
+    it "assigns a new User to @user" do
+      get :new
+      expect(assigns(:user)).to be_a_new(User)
+    end
+
+    it "renders the new template" do
+      get :new
+      expect(response).to render_template(:new)
+    end
+
+    it "returns a 200 status code" do
+      get :new
+      expect(response).to have_http_status(:ok)
+    end
+  end
+
+  describe "POST #create" do
+    let(:valid_params) { attributes_for(:user) }
+    let(:invalid_params) { attributes_for(:user, username: "") }
+
+    context "with valid attributes" do
+      it "creates a new user" do
+        expect {
+          post :create, params: { user: valid_params }
+        }.to change(User, :count).by(1)
+      end
+
+      it "resets the session" do
+        expect(controller).to receive(:reset_session)
+        post :create, params: { user: valid_params }
+      end
+
+      it "logs in the user" do
+        post :create, params: { user: valid_params }
+        expect(session[:user_id]).to eq(User.last.id)
+      end
+
+      it "sets the success flash message" do
+        post :create, params: { user: valid_params }
+        expect(flash[:success]).to eq(I18n.t("users.create.success"))
+      end
+
+      it "redirects to root path" do
+        post :create, params: { user: valid_params }
+        expect(response).to redirect_to(root_path)
+      end
+    end
+
+    context "with invalid attributes" do
+      it "does not create a new user" do
+        expect {
+          post :create, params: { user: invalid_params }
+        }.not_to change(User, :count)
+      end
+
+      it "renders the new template with unprocessable_entity status" do
+        post :create, params: { user: invalid_params }
+        expect(response).to render_template(:new)
+        expect(response).to have_http_status(:unprocessable_entity)
+      end
+    end
+  end
+
+  describe "GET #show" do
+    let!(:user) { create(:user) }
+    let!(:post1) { create(:post, user: user, created_at: 1.day.ago) }
+    let!(:post2) { create(:post, user: user, created_at: 2.days.ago) }
+
+    context "when user is found" do
```

Comment:
context, it, describer dùng khi nào nhỉ

Reply:
- describe: Mô tả nhóm các test case liên quan.
- context: Nhóm các test case theo các điều kiện cụ thể.
- it: Mô tả một test case cụ thể.